### DISCUSSION TITLE
Visual-only guidance and landing with Matek F411-Wing

### DISCUSSION THREAD

**Discussion by juju988** [iNavFlight/inav#6508] [Q&A] [answered]
Hi folks - I'm working on a university project where the plan is to guide a glider to a landing zone using visual guidance only (no gps). We have a Matek F411-wing for control. What I am hoping to do is have the F411 keep the plane stable while a raspberry pi works out the relative co-ordinates and flight path to the targeted landing point. The landing point will be a big "fiducial marker" (like a simple QR code).

My question is - could I grab some thoughts about the easiest way to send control inputs into the F411? My first thought was for the raspberry pi to create GPS-type data based on where it thinks it is, and just pretend to be a GPS sensor.

Another way might be to have the raspberry pi pretend to be an RC receiver and send in commands up/down/left/right commands.

All thoughts appreciated.

Thanks

**Comment by juju988** [comment]
... just to clarify, the raspberry pi is onboard the glider along with the F411

**Comment by juju988** [comment]
Following up my own answer - NMEA sentences from raspberry pi over serial to UART1 looks like it might be feasible. 
Still would appreciate any comments/ideas.

**Reply by jasonleekennedy** [reply]
I would be curious if your 'faked' GPS data could be stable enough to keep the auto pilot code from freaking out (other sensors in conflict for example or air speed calculations/etc).

My first though for something like this would be to put the iNav system in angle mode, then have the raspberry pi send inputs as if they came from a TX system (you would want to connect the RX to the pi, so you have a way to take manual control if something goes haywire).

**Reply by juju988** [reply]
Hi thanks so much for taking the time to reply. 

I agree with you. Looks like I can send commands from the Pi to the FC via SBUS. The FC can limit the elevator/aileron movements to prevent them going too far.

Due to wing size limitations the glider will have a glide slope of a brick (actually similar to the space shuttle) so it will need a bit of a dive first to have enough speed for a flared landing. For this reason I think "direct" control might be better than the NMEA GPS/RTH method that I had been originally thinking of.

Thanks again

**Comment by danarrib** [comment]
Faking a GPS would be doable, but also super difficult.
Enabling ANGLE mode and making RPi pilot the aircraft manually would be a lot easier.
You just have to add an Airspeed sensor, so you can program a consistent glide slope to avoid stall.

You can send RC channels to the Flight Controller over any supported serial protocol (like SBUS). But it would be easier if you do it over MSP, because with a single serial connection you can send RC channel values and fetch flight information from INAV... And it would keep one free UART for a real receiver.

**Reply by juju988** [reply]
Hi thanks a lot for your reply. 

I confirm we're going for the ANGLE mode now per jasonleekennedy, and using SBUS from receiver to the flight controller (we're using a FlySky FS-i6X transmitter and FSIA6B receiver). 

We are currently looking at using a Polulu RC digital switch than can be turned on/off using one of the switches on the transmitter. The RC digital switch (and a transistor) will control the input to two AND gates which I am hoping will give us the ability to switch the serial data going to the flight controlelr from one source to the other (effectively auto-mode from the Raspberry Pi, and manual-override mode directly from the receiver). Just waiting for some bits to arrive from ebay.

If there is an easier way to do it I would love to know!
:)

Thanks again

**Reply by danarrib** [reply]
Maybe it'll be easier if you wire the receiver to the RPi and then the RPi to the FC.
And on your program running on the RPi, you decide if the program will pilot the aircraft or just pass thru the RC channels values from the receiver to the FC.

Less parts involved, less failure points.

**Reply by juju988** [reply]
Thanks danarrib, that was our initial design, but there is a need for redundancy for the competition and we needed to be able to maintain control even if the Raspberry Pi froze and the glider was flying towards something expensive.
:)

**Comment by juju988** [comment, marked as answer]
Thank you very much to @jasonleekennedy and @danarrib for their input. I am marking this as answered.
